lib/collection/src/shards/replica_set/vector_clock.rs

```diff
@@ -0,0 +1,81 @@
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+
+use crate::operations::clock_sync::ClockSync;
+use crate::shards::shard::PeerId;
+
+pub struct VectorClock {
+    counters: Vec<Arc<AtomicU64>>,
+    availabilities: Vec<Arc<AtomicBool>>,
+}
+
+pub struct ClockGuard {
+    id: usize,
+    counter: Arc<AtomicU64>,
+    availability: Arc<AtomicBool>,
+}
+
+impl ClockGuard {
+    pub fn new(id: usize, counter: Arc<AtomicU64>, availability: Arc<AtomicBool>) -> Self {
+        Self {
+            id,
+            counter,
+            availability,
+        }
+    }
+
+    pub fn get_id(&self) -> usize {
+        self.id
+    }
+
+    pub fn resolve(self) {
+        self.availability.swap(false, Ordering::Relaxed);
+    }
+
+    pub fn get_clock_sync(&self, peer_id: PeerId) -> ClockSync {
+        ClockSync {
+            tick: self.counter.load(Ordering::Relaxed),
+            peer_id,
+            thread_id: self.id as u64,
+        }
+    }
```

Comment:
I think `peer_id` should be set once during `VectorClock` construction. Way too easy to accidentally pass wrong `peer_id` into `get_clock_sync`.